### PR TITLE
Integration specs

### PR DESCRIPTION
**Pull request by AlierPM**
# Blog App - Integration Specs for Views and Fix N+1 Problems
### Project Submission Status

- Submitted this project as a solo submission. My coding partner faced issues. With PERMISSION FROM THE STUDENT SUCCESS TEAM, and I am providing the screenshot of the email in this PR.
Screenshot from 2024-01-25 17-06-43

## Learning Objectives

- Wrote integration tests with the Capybara gem.
- Secured the app from N+1 problems.

## Time Spent

Approximately 4.5 hours.

## Description

In this project, integration tests were created for all views in the app. User stories or workflows were covered, considering possible errors users may cause.

**IMPORTANT NOTE:** Please review all requirements to ensure completion.

## General Requirements

- Ensured there are no linter errors.
- Used correct Gitflow.
- Documented work professionally.

## Ruby Requirements

- Followed best practices for Ruby.

### N+1

- Solved the N+1 problem when fetching all posts and their comments for a user.
- Did not use the bullet gem in this exercise; analyzed the logs as shown in [this article](link-to-article).
- Included screenshots of the app's logs before and after the fix in the PR description.
Screenshot from 2024-01-26 21-07-45

Screenshot from 2024-01-26 21-20-04

### Integration Specs

Used Capybara to write integration tests for each view in the project.

#### User Index Page

- Could see the username of all other users.
- Could see the profile picture for each user.
- Could see the number of posts each user has written.
- Clicking on a user redirected to that user's show page.

#### User Show Page

- Could see the user's profile picture.
- Could see the user's username.
- Could see the number of posts the user has written.
- Could see the user's bio.
- Could see the user's first 3 posts.
- Could see a button that lets me view all of a user's posts.
- Clicking a user's post redirected to that post's show page.
- Clicking to see all posts redirected to the user's post index page.

#### User Post Index Page

- Could see the user's profile picture.
- Could see the user's username.
- Could see the number of posts the user has written.
- Could see a post's title.
- Could see some of the post's body.
- Could see the first comments on a post.
- Could see how many comments a post has.
- Could see how many likes a post has.
- Could see a section for pagination if there are more posts than fit on the view.
- Clicking on a post redirected to that post's show page.

#### Post Show Page

- Could see the post's title.
- Could see who wrote the post.
- Could see how many comments it has.
- Could see how many likes it has.
- Could see the post body.
- Could see the username of each commentor.
- Could see the comment each commentor left.

### CODE REVIEW

File: spec/features/user_show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    expect(page).to have_link('See all posts')
+  end
+
+  it 'redirects to user post\'s index page when "See all posts" is clicked' do
+    visit user_path(@user)
+    click_link 'See all posts'
+    sleep 15
+    expect(current_path).to eq(user_posts_path(@user))
+  end
+end
```
Comment (TedLivist) [inline spec/features/user_show_spec.rb]: - [x] Kindly add the missing test:
  
  - Tests that asserts you can see the user's first 3 posts.

File: spec/features/user_post_index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  end
+
+  it 'redirects to post show page when clicking on a post' do
+    visit user_posts_path(@user)
+
+    # Click the first post's title
+    click_link @posts.first.title
+
+    expect(current_path).to eq(user_post_path(@user, @posts.first))
+  end
```
Comment (TedLivist) [inline spec/features/user_post_index_spec.rb]: - [x] Kindly add the missing test:

  - Tests that asserts you can see the first comments on a post.

**Review by TedLivist — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ ♻️

Hi @AlierPM,

Good job so far!
✔️ Well-written and passing tests
However, there are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/user_show_spec.rb (hunk 1 above)
Comment (AlierPM) [reply spec/features/user_show_spec.rb]: Thanks for the constructive feedback; well perceived and done.

**Review by TedLivist — APPROVED** [review]
## STATUS: APPROVED ✔️ ✔️

Hi @AlierPM,

✔️ Well-written and passing tests

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**